File: bench/scala/src/main/scala/ming/TestRunner.scala
package ming

import com.google.gson.{Gson, JsonArray, JsonObject}

import java.nio.file.{Files, Path}
import scala.jdk.CollectionConverters.*

/** Standalone test runner — no test framework needed.
  *
  * Reads tests.json + fixtures, calls evalStr/evalStrWithOutput, compares results. Exit 0 = all pass, exit 1 = failures
  * found.
  *
  * Usage: java -jar test.jar 01 # run level 1 java -jar test.jar all # run all levels
  *
  * Env vars: TESTS_JSON — path to tests.json (default ../tests.json) FIXTURES_DIR — path to fixtures dir (default
  * ../fixtures)
  */
object TestRunner:

  private case class TestResult(name: String, passed: Boolean, message: String)

  def main(args: Array[String]): Unit =
    val levelArg = args.headOption.getOrElse("all")
    val benchLevel = levelArg match
      case "all" => 0
      case s     => s.toIntOption.getOrElse(0)
    System.setProperty("bench.level", benchLevel.toString)

    val testsJsonPath = Path.of(
      Option(System.getenv("TESTS_JSON")).getOrElse("../tests.json")
    )
    val fixturesDir = Path.of(
      Option(System.getenv("FIXTURES_DIR")).getOrElse("../fixtures")
    )

    val testCases: JsonArray =
      val json = Files.readString(testsJsonPath)
      Gson().fromJson(json, classOf[JsonArray])

    val results: List[TestResult] =
      testCases.iterator().asScala.toList.flatMap { elem =>
        val tc      = elem.getAsJsonObject
        val name    = tc.get("name").getAsString
        val level   = tc.get("level").getAsInt
        val fixture = tc.get("fixture").getAsString
        val kind    = tc.get("kind").getAsString

        val skip =
          (benchLevel > 0 && level > benchLevel) ||
            (tc.has("deprecated_after") && benchLevel > 0 &&
              benchLevel > tc.get("deprecated_after").getAsInt)

        if skip then None
        else
          val input = Files.readString(fixturesDir.resolve(fixture))
          Some(runTest(name, kind, input, tc))
      }

    results.foreach { r =>
      if r.passed then println(s"PASS ${r.name}")
      else println(s"FAIL ${r.name}: ${r.message}")
    }

    val passed = results.count(_.passed)
    val failed = results.count(!_.passed)
    val total  = results.size
    println(s"$passed passed, $failed failed out of $total tests")

    if failed > 0 then System.exit(1)

  private def runTest(
    name: String,
    kind: String,
    input: String,
    tc: JsonObject
  ): TestResult =
    try
      kind match
        case "eval_str_ok" =>
          val expected = tc.get("expected").getAsString
          val result   = evalInput(input, tc)
          if result == expected then TestResult(name, passed = true, "")
          else TestResult(name, passed = false, s"expected $expected got $result")

        case "eval_str_err" =>
          try
            val result = evalInput(input, tc)
            TestResult(name, passed = false, s"expected EvalError but got $result")
          catch case _: EvalError => TestResult(name, passed = true, "")

        case "eval_str_err_with_position" =>
          try
            val result = evalInput(input, tc)
            TestResult(name, passed = false, s"expected EvalError but got $result")
          catch
            case e: EvalError =>
              if e.getMessage.matches(".*\\d+:\\d+.*") then TestResult(name, passed = true, "")
              else
                TestResult(
                  name,
                  passed = false,
                  s"error message should contain line:col position, got: ${e.getMessage}"
                )

        case "eval_str_with_output" =>
          val expectedOutput = tc.get("expected_output").getAsString
          val (_, output)    = Evaluator.evalStrWithOutput(input)
          if output == expectedOutput then TestResult(name, passed = true, "")
          else TestResult(name, passed = false, s"expected output $expectedOutput got $output")

        case other =>
          TestResult(name, passed = false, s"unknown test kind: $other")
    catch
      case e: Exception =>
        TestResult(name, passed = false, s"${e.getClass.getSimpleName}: ${e.getMessage}")

  private def evalInput(input: String, tc: JsonObject): String =
    Option.when(tc.has("max_steps"))(tc.get("max_steps").getAsInt) match
      case Some(maxSteps) =>
        Evaluator.evalStrWithLimit(input, maxSteps)
      case None =>
        Evaluator.evalStr(input)

File: bench/java/src/main/java/ming/TestRunner.java
package ming;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class TestRunner {

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: java ming.TestRunner <level|all>");
            System.exit(2);
        }

        String levelArg = args[0];
        int benchLevel = 0;
        if (!levelArg.equals("all")) {
            try {
                benchLevel = Integer.parseInt(levelArg);
            } catch (NumberFormatException e) {
                System.err.println("Invalid level: " + levelArg);
                System.exit(2);
            }
        }

        String testsJsonPath = System.getenv("TESTS_JSON");
        if (testsJsonPath == null || testsJsonPath.isEmpty()) {
            testsJsonPath = "../tests.json";
        }
        String fixturesDir = System.getenv("FIXTURES_DIR");
        if (fixturesDir == null || fixturesDir.isEmpty()) {
            fixturesDir = "../fixtures";
        }

        Path testsJson = Path.of(testsJsonPath);
        Path fixtures = Path.of(fixturesDir);

        JsonArray testCases;
        try {
            String json = Files.readString(testsJson);
            testCases = new Gson().fromJson(json, JsonArray.class);
        } catch (IOException e) {
            System.err.println("Failed to load tests.json from " + testsJson + ": " + e.getMessage());
            System.exit(2);
            return;
        }

        int passed = 0;
        int failed = 0;
        int total = 0;

        for (JsonElement elem : testCases) {
            JsonObject tc = elem.getAsJsonObject();
            String name = tc.get("name").getAsString();
            int level = tc.get("level").getAsInt();
            String fixture = tc.get("fixture").getAsString();
            String kind = tc.get("kind").getAsString();

            if (benchLevel > 0 && level > benchLevel) {
                continue;
            }

            if (tc.has("deprecated_after") && benchLevel > 0) {
                int deprecatedAfter = tc.get("deprecated_after").getAsInt();
                if (benchLevel > deprecatedAfter) {
                    continue;
                }
            }

            String testName = "test_" + name;
            total++;

            try {
                String input = Files.readString(fixtures.resolve(fixture));
                Evaluator evaluator = new Evaluator();

                switch (kind) {
                    case "eval_str_ok" -> {
                        String expected = tc.get("expected").getAsString();
                        String result = evaluator.evalStr(input);
                        if (expected.equals(result)) {
                            System.out.println("PASS " + testName);
                            passed++;
                        } else {
                            System.out.println("FAIL " + testName + ": expected " + expected + " got " + result);
                            failed++;
                        }
                    }
                    case "eval_str_err" -> {
                        try {
                            String result = evaluator.evalStr(input);
                            System.out.println("FAIL " + testName + ": expected EvalError but got " + result);
                            failed++;
                        } catch (EvalError e) {
                            System.out.println("PASS " + testName);
                            passed++;
                        }
                    }
                    case "eval_str_err_with_position" -> {
                        try {
                            String result = evaluator.evalStr(input);
                            System.out.println("FAIL " + testName + ": expected EvalError but got " + result);
                            failed++;
                        } catch (EvalError e) {
                            if (e.getMessage() != null && e.getMessage().matches(".*\\d+:\\d+.*")) {
                                System.out.println("PASS " + testName);
                                passed++;
                            } else {
                                System.out.println("FAIL " + testName + ": error message should contain line:col position, got: " + e.getMessage());
                                failed++;
                            }
                        }
                    }
                    case "eval_str_with_output" -> {
                        String expectedOutput = tc.get("expected_output").getAsString();
                        EvalResult result = evaluator.evalStrWithOutput(input);
                        if (expectedOutput.equals(result.output())) {
                            System.out.println("PASS " + testName);
                            passed++;
                        } else {
                            System.out.println("FAIL " + testName + ": expected output " + expectedOutput + " got " + result.output());
                            failed++;
                        }
                    }
                    default -> {
                        System.out.println("FAIL " + testName + ": unknown test kind: " + kind);
                        failed++;
                    }
                }
            } catch (EvalError e) {
                System.out.println("FAIL " + testName + ": unexpected EvalError: " + e.getMessage());
                failed++;
            } catch (IOException e) {
                System.out.println("FAIL " + testName + ": failed to read fixture: " + e.getMessage());
                failed++;
            } catch (Exception e) {
                System.out.println("FAIL " + testName + ": " + e.getClass().getSimpleName() + ": " + e.getMessage());
                failed++;
            }
        }

        System.out.println(passed + " passed, " + failed + " failed out of " + total + " tests");

        // Run L27+ standalone tests if level includes them
        boolean extraFailed = false;
        if (benchLevel == 0 || benchLevel >= 27) {
            extraFailed |= runStandaloneTests("ming.L27Tests");
        }
        if (benchLevel == 0 || benchLevel >= 28) {
            extraFailed |= runStandaloneTests("ming.L28Tests");
        }

        System.exit((failed > 0 || extraFailed) ? 1 : 0);
    }

    private static boolean runStandaloneTests(String className) {
        try {
            Class<?> cls = Class.forName(className);
            // Reset pass/fail counters
            var passedField = cls.getDeclaredField("passed");
            var failedField = cls.getDeclaredField("failed");
            passedField.setInt(null, 0);
            failedField.setInt(null, 0);
            // Invoke each test method (static void testXxx) directly, bypassing main/System.exit
            for (var m : cls.getDeclaredMethods()) {
                if (m.getName().startsWith("test") && m.getParameterCount() == 0) {
                    m.setAccessible(true);
                    m.invoke(null);
                }
            }
            int p = passedField.getInt(null);
            int f = failedField.getInt(null);
            System.out.println(p + " passed, " + f + " failed out of " + (p + f) + " tests");
            return f > 0;
        } catch (ClassNotFoundException e) {
            return false; // not present — skip
        } catch (Exception e) {
            System.out.println("FAIL " + className + ": " + e);
            return true; // failure
        }
    }
}
